metadata/src/main/java/org/apache/kafka/controller/ClusterControlManager.java

```diff
@@ -284,7 +286,7 @@ ReplicaPlacer replicaPlacer() {
     public void activate() {
         heartbeatManager = new BrokerHeartbeatManager(logContext, time, sessionTimeoutNs);
         for (BrokerRegistration registration : brokerRegistrations.values()) {
-            heartbeatManager.touch(registration.id(), registration.fenced(), -1);
+            heartbeatManager.register(registration.id(), registration.fenced(), registration.directories());
```

Comment:
This change seems a bit odd to me.  I think it will work, and in fact I think it wil do exactly what was being done before, but you moved the creation of the broker entry in the heartbeat manager's internal map into the `register()` method and removed it from the `touch()` method -- which I think is a god change.  But now this reads weirdly because we aren't registering a broker -- the broker should already have been registered by this point, and we just want to "touch" it.  I think this change can be reverted?

Reply:
This is in `clusterControlManager.activate()`, a couple of lines above we reset the heartbeatManager with `heartbeatManager = new BrokerHeartbeatManager(logContext, time, sessionTimeoutNs);`. So no broker will be registered yet with that fresh manager.

`register()` here means _register the broker in the heartbeat manager_, not _register the broker with the controller quorum_. 

If we revert this line change we'll trigger `throw new IllegalStateException("Broker " + brokerId + " is not registered.");` in `brokerHeartbeatManager.touch()`.

Reply:
Ah, okay, yes, makes sense.  Thanks.

Reply:
I don't see why we need to handle directories in `HeartbeatManager`. I made this comment in the other PR too. HBM is about timing out brokers. It's not about JBOD stuff really.

Directories are hard state so they should be kept in `ClusterControlManager`. And we need to issue a new `BrokerRegistrationRecord` when they change.

Reply:
> And we need to issue a new BrokerRegistrationRecord when they change.
We do that in #14902 which builds on this.

I have removed directory state from the heartbeat manager, we now rely only on `ClusterControlManager`, and any persisted metadata state.